infer/src/backend/dom.ml

```diff
@@ -2169,55 +2169,145 @@ let pathset_join pname tenv (pset1: Paths.PathSet.t) (pset2: Paths.PathSet.t)
+  let module S = MaxCliqueState in
+  (S.add first S.empty, n, edge)
+
+(* Given a maxclique generator, returns a pair (xs, g) with the next maxclique
+[xs] and the updated generator state [g]. If no more maxclique, raises
+[Not_found]. For the algorithm, see the JYP paper mentioned below. *)
+let next_maxclique (queue, n, edge) =
+  let module S = MaxCliqueState in
+  let xs = S.min_elt queue in (* raises [Not_found] at the end *)
+  let queue = S.remove xs queue in
+  let ys =
+    let ok_yx y x = x < y && not (edge x y) in
+    let ok_y y _ = List.exists xs ~f:(ok_yx y) in
+    filter_to n ok_y [] 0 in
+  let do_y queue y =
+    let xy_compat x = x < y && edge x y in
+    let zs = List.filter xs ~f:xy_compat @ [y] in
+    let maxclique zs = (* is zs a maxclique for the subgraph induced by 0..y? *)
+      let rec loop u vs = u > y || (match u, vs with
+        | u, v :: vs when Int.equal u v -> loop (u + 1) vs
+        | u, vs -> not (List.for_all zs ~f:(edge u)) && loop (u + 1) vs) in
+      loop 0 zs in
+    let fillup zs = (* greedily add vertices from y+1..n-1 to clique *)
+      let compatible z zs = List.for_all zs ~f:(edge z) in
+      filter_to n compatible (List.rev zs) (y + 1) in
+    if maxclique zs then S.add (fillup zs) queue else queue in
+  let queue = List.fold ~init:queue ~f:do_y ys in
+  (xs, (queue, n, edge))
+
+(* Find maximal subsets of [xs] that can be combined with a binary operator
+[plus] that is associative, commutative, but not necessarily defined
+everywhere.
+
+Let's call set S={x1,...,xn} *consistent* when ΣS=x1+...+xn is defined.  When
+x1,...,xn are preconditions and [plus] is the meet operator, it happens often
+(but not always) that pairwise consistency implies set consistency.  Therefore,
+we will use an algorithm that works fast in this case, and is incomplete
+otherwise.
+
+Suppose pairwise consistency implies set consistency.  Define a graph whose
+vertices are the [xs] and has an (undirected) edge xy when x+y is defined. In
+terms of this graph, the problem is to generate all maximal cliques. This can
+be done with polynomial delay using the JYP algorithm:
+  Johnson, Yannakakkis, Papadimitriou,
+  On Generating All Maximal Independent Sets, 1987
+A maximal clique S found by this algorithm will be pairwise consistent, and
+often but not always consistent. The current implementation simply filters
+out inconsistent maximal cliques.
+
+The [timeout] limits the number of maxcliques to be enumerated, because they
+can be exponentially many.
+*)
+let maximal_combine ~(timeout : int) ~(init : 'a) ~(plus : 'a -> 'a -> 'a option) (xs : 'a list)  : 'a list =
+  let n = List.length xs in
+  let by_idx =
+    let xs = Array.of_list xs in
+    Array.get xs in
+  let lost = ref 0 in (* counts inconsistent maximal cliques *)
+  let exception Undefined in
+  let plus_exc x y = match plus x y with
+    | None -> raise Undefined
+    | Some z -> z in
+  let pairs = (* first, build graph with pairwise consistency *)
```

Comment:
Please have `pairs` defined in `edge`, like you did with `by_idx`

Reply:
This became the `cache` now, and it is inside `plus_by_idx`.